documents/final-submission-document/chapters/3_concept.tex

```diff
@@ -62,26 +62,35 @@ \subsection{Web Application Topology}
 To manage the Controller from our interface, we considered the implementation of either directly connecting to the XMPP server over WebSockets~\cite{rfc7395} or HTTP (BOSH~\cite{xep-0124}), or to communicate indirectly with the XMPP server via custom WebAPI Proxy.
 These topologies are illustrated in Figure~\ref{fig:architecturecontainerwebapplication} and Figure~\ref{fig:architecturecontainerwebproxy}.
 
-As elaborated in the according design decision (see Appendix~\ref{sec:architectural-decisions}), we decided for the direct connection via WebSockets, with a fallback to BOSH if the required features are not supported (e.g. SASL EXTERNAL authentication).
-This topology simplifies the implementation and deployment of the application in comparison to a WebAPI Proxy.
-WebSockets offer stateful TCP-sockets to exchange data with the XMPP server in contrast to BOSH, which uses HTTP long polling to emulate a bidirectional stream and is therefore less efficient~\cite{xep-0124}.
-
-Using the XMPP Service Discovery~\cite{xep-0030}, the XMPP server may be queried for supported features, so that only supported functionality is presented to the application user.
-
 \begin{figure}[h]
 \centering
 \includegraphics[width=0.7\linewidth]{resources/architecture_container_webapplication}
 \caption[Architecture Container Diagram: Web Application]{Architecture Container Diagram showing the web application topology with WebSockets or BOSH.}
 \label{fig:architecturecontainerwebapplication}
 \end{figure}
 
+\subsubsection{WebAPI Proxy}
+
+A WebAPI Proxy could be realised with a custom browser-to-proxy protocol, as implemented in the XMPP-FTW JavaScript library\footnote{\url{http://docs.xmpp-ftw.org/}}.
+However, this approach leads to a high coupling between a concrete library and the web application.
+
+Another approach would be the implementation of a custom WebSocket-to-XMPP Proxy, which allows connecting to XMPP servers that do not fully support WebSockets or BOSH.
+If such a proxy is implemented transparently, the client is not aware of the server limitations. Therefore, the client implementation is no different from direct communication with an XMPP server.
+
 \begin{figure}[h]
 \centering
-\includegraphics[width=0.7\linewidth]{resources/architecture_container_proxy.pdf}
+\includegraphics[width=\linewidth]{resources/architecture_container_proxy.pdf}
 \caption[Architecture Container Diagram: Web Proxy]{Architecture Container Diagram showing the WebAPI Proxy topology.}
 \label{fig:architecturecontainerwebproxy}
 \end{figure}
 
+\subsubsection{Implemented Web Application Topology}
+
+As elaborated in the according design decision (see Appendix~\ref{sec:architectural-decisions}), we decided for the direct connection via WebSockets, with a fallback to BOSH if the required features are not supported (e.g. SASL EXTERNAL authentication).
```

Comment:
I think this section is a bit confusing as some concepts are mixed. BOSH is a fallback for WebSockets but not (necessarily) a fallback for missing authentication features. Also, if it were a fallback for authentication features, this relation must be explained.